### ISSUE TITLE
process_spider_output called twice when exception occurs

### ISSUE THREAD

**Issue by StasDeep** [scrapy/scrapy#4266]
### Description

Don't know if this should be considered a bug or not, but it looks very unintuitive.

When raising an exception in a generator callback, the `process_spider_output` method of a spider middleware, that goes after a middleware that catches an exception, is invoked twice. Once for the result of the callback that was yielded before exception occurred and one more time for the iterable returned by the `process_spider_exception` method of the middleware that caught an exception.

I think it may make more sense to pass objects yielded by the callback to the `process_spider_exception` so that middleware can have control over what to do with objects yielded by a callback that raised an exception. Or, we should have a way of checking if the object passed to `process_spider_output` was generated by a callback that eventually raised an exception. For example, I want to modify an item generated by a callback if there was an exception after it was generated.

At the very least, this should be documented to not mislead.

### Code to Reproduce

```python
import scrapy


class FirstExceptionHandlerMiddleware:
    def process_spider_exception(self, response, exception, spider):
        print('Exception caught with First:', exception)
        return []

    def process_spider_output(self, response, result, spider):
        result = list(result)
        print('Processing spider output with First:', result)
        yield from result


class SecondExceptionHandlerMiddleware:
    def process_spider_output(self, response, result, spider):
        result = list(result)
        print('Processing spider output with Second:', result)
        yield from result


class ExceptionSpider(scrapy.Spider):
    name = 'exception_spider'
    start_urls = ['https://example.org']
    custom_settings = {
        'SPIDER_MIDDLEWARES': {
            __name__ + '.FirstExceptionHandlerMiddleware': 870,
            __name__ + '.SecondExceptionHandlerMiddleware': 860,
        }
    }

    def parse(self, response):
        yield {'an': 'item'}
        raise Exception('foo')
```

### Output

```
Exception caught with First: foo
Processing spider output with First: [{'an': 'item'}]
Processing spider output with Second: [{'an': 'item'}]
Processing spider output with Second: []
```

### Versions

```
Scrapy       : 1.7.1
lxml         : 4.2.5.0
libxml2      : 2.9.8
cssselect    : 1.0.3
parsel       : 1.5.1
w3lib        : 1.19.0
Twisted      : 18.9.0
Python       : 3.6.8 (default, Sep  5 2019, 08:52:26) - [GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.10.44.4)]
pyOpenSSL    : 18.0.0 (OpenSSL 1.1.0j  20 Nov 2018)
cryptography : 2.4.2
Platform     : Darwin-19.2.0-x86_64-i386-64bit
```

**Comment by Gallaecio** [comment]
Related to #4260

**Comment by elacuesta** [comment]
I don't think this is a bug, I think it's just not properly documented, just like the fact that the `process_spider_exception` chain is invoked parallel to the `process_spider_output` one.

This is what's happening, ignore my poor image editing skills:
![spider_middleware](https://user-images.githubusercontent.com/1731933/71917975-1a888f00-3160-11ea-9fb2-2a904967ff2d.jpg)

I will try to write some docs for this section, and turn the above drawing into a relatively good-looking diagram.

I am going to take the liberty of removing the "bug" label, but feel to re-apply if you think it's necessary, and of course let's keep the discussion open.

**Comment by elacuesta** [comment]
Regarding the proposed solutions:

> I think it may make more sense to pass objects yielded by the callback to the `process_spider_exception` so that middleware can have control over what to do with objects yielded by a callback that raised an exception

I think this would imply a rewrite of the spider component, which is likely to require some additional discussion.

> Or, we should have a way of checking if the object passed to `process_spider_output` was generated by a callback that eventually raised an exception

This sounds feasible, maybe adding a flag attribute to the iterable, indicating its origin.

**Comment by AdrianAtZyte** [comment]
> This is stale. The pipeline that produced this got rewritten (async generators, [#4260](https://github.com/scrapy/scrapy/pull/4260) and later), and the current design no longer replays an item through a middleware’s `process_spider_output` that it already went through: items recovered from `process_spider_exception` only re-enter `process_spider_output` for the middlewares *after* the one that caught the exception, see [`spidermw.py#L142`](https://github.com/scrapy/scrapy/blob/da7c2fc59e286e796d0f8cfe01d4363d6fa3729f/scrapy/core/spidermw.py#L142). This exact scenario is pinned by [`test_generator_output_chain`](https://github.com/scrapy/scrapy/blob/da7c2fc59e286e796d0f8cfe01d4363d6fa3729f/tests/test_spidermiddleware_output_chain.py#L393), and the hand-off behavior is documented in the [spider middleware docs](https://docs.scrapy.org/en/latest/topics/spider-middleware.html#scrapy.spidermiddlewares.SpiderMiddleware.process_spider_exception).
